cmd/kubeadm/app/cmd/completion_test.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"github.com/spf13/cobra"
+	"testing"
+)
+
+func TestNewCmdCompletion(t *testing.T) {
+	var out bytes.Buffer
+	shells := GetSupportedShells()
+	if len(shells) == 0 {
+		t.Errorf("GetSupportedShells() returned zero")
+	}
+	// test NewCmdCompletion with a valid shell.
+	// use a dummy parent command as NewCmdCompletion needs it.
+	parentCmd := &cobra.Command{}
+	args := []string{"completion", shells[0]}
+	parentCmd.SetArgs(args)
+	cmd := NewCmdCompletion(&out, "")
+	parentCmd.AddCommand(cmd)
+	if err := parentCmd.Execute(); err != nil {
+		t.Errorf("Cannot exectute NewCmdCompletion: %v", err)
+	}
+}
+
+func TestRunCompletion(t *testing.T) {
```

Comment:
IMO this test can be merged into above test case.

Reply:
but then the test would exit if the command fails (see `kubeadmutil.CheckErr(err)`) and we won't be able to test all the test cases with `expectedError: true,`.

Reply:
> we won't be able to test all the test cases with expectedError: true

@neolit123 You can use `recover()` to cover this test scenario. Please refer to https://github.com/kubernetes/kubernetes/blob/36058cb0c3ed3c9e46e1eda48c106679b603d0cb/pkg/kubelet/cm/cpumanager/policy_static_test.go#L91-L101 for guidance.

Reply:
thank you, i will try `recover()`.

Reply:
@dixudx 

i think you suggestion to use `recover()` would not work in this case.
please have a look here, where `CheckErr()` is used to validate if `RunComletion()` has error:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/completion.go#L104
`CheckErr()` calls `checkErr("", err, fatal)`, where `fatal()` is a callback that calls `os.Exit(code)`
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/error.go#L59

correct me if i'm wrong, but `recover()` would not recover from an `os.Exit()` because the whole process is terminated and this is not a `panic`?

Reply:
@neolit123 Yes, you're right. My bad. Sorry for the misleading.

So keep it as it is.